### ISSUE TITLE
Maxwell Forceful Shutdown 

### ISSUE THREAD

**Issue by shannontan-addepar** [zendesk/maxwell#1046]
Maxwell shutdowns forcefully when a specific thread doesn't stop within the specified timeout. It calls `Runtime.halt` and not `System.exit`. In certain scenarios where Maxwell is running embedded in an application, this can produce issues. 

Is this required? Can a graceful shutdown be satisfactory? I can't think of any scenarios where the forceful shutdown is required, but I defer to your expertise.

**Comment by shannontan-addepar** [comment]
Alternatively, should we add functionality to specify exit codes for the purpose of post-exit behavior, such as by JSVC?

**Comment by timbertson** [comment]
Yeah, unfortunately this has been necessary in the past. To shut down gracefully without this, you would need to ensure:

 - you only have one non-daemon thread, and
 - that thread ends whenever any fatal event happens (explicit shutdown, uncaught exception, etc)

Libraries can and do run non-daemon threads, so (1) is out. And we've had experiences in the past where things that should work on shutdown would actually stall the system and _prevent_ shutdown (e.g. kafka's `producer.close()` when it's in a bad state). So I'd love to support more graceful shutdown, but the aggressive shutdown mechanisms were added out of necessity.

Embedded maxwell is an obvious case where this _might_ not be appropriate, but making that configurable could get messy. I guess there could be some kind of shutdownHook which you can override in embedded mode if you want something less aggressive, but I'm just thinking out loud (I haven't thought much about what the interface would look like).

I'm not sure what you mean by specifying exit codes, do you want your code to be able tell embedded maxwell to shutdown with a certain exit code?

**Comment by osheroff** [comment]
@timbertson do you remember what the nature of the stalling errors was?  Was the thread totally non-responsive or did it throw an exception on close?

**Comment by timbertson** [comment]
That one was nonresponsive - it tried to close the connection, but then it waited for some kind of completion which never came (I guess the underlying connection was already in a bad state). No exceptions or anything, just silence 🙉

**Comment by shannontan-addepar** [comment]
Hmm okay - I think these are fair points. At this point, we might leverage alternative restart mechanisms to reduce unnecessary complexity in the process itself (we run it embedded). That said, if others come across this and believe it might have value, I have no opposition to adding it myself. 

Would it be fair to close this out and have others open the ticket if they run into the same issue?

**Comment by osheroff** [comment]
In situtations like this where we have no idea if it's safe to proceed, I think the best thing to do is to throw an exception out of `Maxwell#terminate` instead of calling `System.halt` in the underbellies.  At least then embedded users have a choice of proceeding anyway.
